### PR TITLE
feat: compile dbt project before compile, deploy and preview commands

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #2563

### CODE REVIEW

File: packages/cli/src/handlers/compile.ts [hunk 1]
```diff
@@ -56,8 +61,11 @@ export const compile = async (options: GenerateHandlerOptions) => {
     );
     return explores;
 };
-export const compileHandler = async (options: GenerateHandlerOptions) => {
-    const explores = await compile(options);
+export const compileHandler = async (
+    options: GenerateHandlerOptions,
+    command: any,
+) => {
+    const explores = await compile(options, command);
```
Comment (owlas) [inline packages/cli/src/handlers/compile.ts:68]: Could we add some logging so the user knows that dbt compile is running and expected 

Also how do logs from dbt look? and errors?

File: packages/cli/src/handlers/compile.ts [hunk 2]
```diff
@@ -15,14 +15,19 @@ import {
...
     profilesDir: string;
     target: string | undefined;
     profile: string | undefined;
 };
-export const compile = async (options: GenerateHandlerOptions) => {
+export const compile = async (
+    options: GenerateHandlerOptions,
+    command: any,
+) => {
+    await dbtCompile(options, command.parent.args.slice(1));
```
Comment (owlas) [inline packages/cli/src/handlers/compile.ts:30]: why do we need this `command` ? Aren't all the options for `dbt compile` in the `options` object?

**Review by owlas — COMMENTED** [review]
couple questions!

File: packages/cli/src/handlers/dbt/compile.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    vars: string | undefined;
+    threads: string | undefined;
+    noVersionCheck: boolean | undefined;
+    exclude: string | undefined;
+    selector: string | undefined;
+    state: string | undefined;
+    fullRefresh: boolean | undefined;
+};
+export const dbtCompile = async (
+    options: DbtCompileOptions,
```
Comment (owlas) [inline packages/cli/src/handlers/dbt/compile.ts:20]: the options aren't used. aren't `args` and `options` the same?

Comment (ZeRego) [reply packages/cli/src/handlers/dbt/compile.ts:20]: Yes. Was trying to keep the code similar to the dbt run command. 
If we use the options object we need to convert the properties from camelcase to kebabcase and prefix `--` which is doable.

File: packages/cli/src/handlers/compile.ts (hunk 1 above)
Comment (ZeRego) [reply packages/cli/src/handlers/compile.ts:68]: I tried to have a spinner but there is already an active spinner for preview/deploy and causes the terminal to flick between the 2.  

I can enable the dbt logs if it does't cause the same issue as above.

Comment (owlas) [reply packages/cli/src/handlers/compile.ts:68]: Yeah I think showing logs is great - just pipe it so it looks identical to if they had run `dbt compile` themselves

File: packages/cli/src/handlers/dbt/compile.ts (hunk 3 above)
Comment (owlas) [reply packages/cli/src/handlers/dbt/compile.ts:20]: Ah yes I see, whatever you think is simpler!

**Review by owlas — APPROVED** [review]

**Pull request merged**